### PR TITLE
more checked arith and other lints

### LINKED ISSUE #2555: Audit for all usage of unchecked arithmetic and update nomenclature
Looking through the codebase for unchecked arithmetic I still find quite a bit, e.g.:
- [here](https://github.com/anoma/namada/blob/4209a7413c21df3f9b1190189990f4146c00dda3/crates/core/src/types/dec.rs#L368)
- [here](https://github.com/anoma/namada/blob/4209a7413c21df3f9b1190189990f4146c00dda3/crates/core/src/types/uint.rs#L667)
- many more, just searching...

We should:
- Remove ALL support for unchecked arithmetic. We should never use it.
- Update all arithmetic to use `checked_sub`, `checked_add` etc. names (don't overload `+` / `-`) to make it clear that we're using checked arithmetic (and so a search for only `+` or `-`, for example, should return only internal implementation code instead decimal types and no usage other than that).

### PR DESCRIPTION
**Pull request by tzemanovic**
## Describe your changes

closes #2555

## Indicate on which release or other PRs this topic is based on

0.35.1

## Checklist before merging to `draft`
- [x] I have added a changelog
- [x] Git history is in acceptable state

### CODE REVIEW

File: crates/governance/src/event.rs [hunk 1]
```diff
@@ -101,14 +102,22 @@ impl GovernanceEvent {
...
 pub enum ProposalEventKind {
     /// New proposal event
-    NewProposal { proposal_type: GovProposalType },
+    NewProposal {
+        /// Type of proposal
+        proposal_type: GovProposalType,
+    },
     /// Passed proposal
     Passed {
+        /// Does the proposal contain code
```
Comment (brentstone) [inline crates/governance/src/event.rs:111]: add a `?` at the end

File: crates/namada/src/ledger/native_vp/ibc/context.rs [hunk 2]
```diff
@@ -80,25 +81,27 @@ where
     fn read_bytes(&self, key: &Key) -> Result<Option<Vec<u8>>> {
         match self.store.get(key) {
             Some(StorageModification::Write { ref value }) => {
-                let gas = key.len() + value.len();
+                let gas = checked!(key.len() + value.len())? as u64;
```
Comment (brentstone) [inline crates/namada/src/ledger/native_vp/ibc/context.rs:84]: fine not to use a `u64::try_from` as in other places? Same for below

File: crates/namada/src/ledger/native_vp/masp.rs [hunk 3]
```diff
@@ -405,9 +405,8 @@ where
             return Err(error);
         }
 
-        let mut transparent_tx_pool = I128Sum::zero();
+        let mut transparent_tx_pool = shielded_tx.sapling_value_balance();
         // The Sapling value balance adds to the transparent tx pool
```
Comment (brentstone) [inline crates/namada/src/ledger/native_vp/masp.rs:409]: put the comment above the code? Or remove?

File: crates/proof_of_stake/src/epoched.rs [hunk 4]
```diff
@@ -121,7 +121,12 @@ where
                                 && epoch
                                     > Self::sub_past_epochs(params, last_update)
                             {
-                                epoch = Epoch(epoch.0 - 1);
+                                epoch = Epoch(
+                                    epoch
+                                        .0
+                                        .checked_sub(1)
+                                        .expect("Cannot underflow"),
```
Comment (brentstone) [inline crates/proof_of_stake/src/epoched.rs:128]: perhaps use `epoch.prev()` instead

File: crates/namada/src/ledger/native_vp/ibc/context.rs (hunk 2 above)
Comment (tzemanovic) [reply crates/namada/src/ledger/native_vp/ibc/context.rs:84]: I think it's ok to assume 64bit arch - we'll have a warning if not (https://github.com/anoma/namada/pull/3215)

Comment (tzemanovic) [reply crates/namada/src/ledger/native_vp/ibc/context.rs:84]: also it won't fail on 32bit

**Comment by tzemanovic** [comment]
> Looks good to me in general, just left a couple questions.

thx! I fixed all the issues in https://github.com/anoma/namada/compare/ec36ae148190e183c70cc0c56171aa3bb8b1d359..31e54b9fcccb58d5628cc0362597d5848b576796

**Pull request merged**